easybuild/tools/filetools.py

```diff
@@ -1476,6 +1480,103 @@ def mkdir(path, parents=False, set_gid=None, sticky=None):
         _log.debug("Not creating existing path %s" % path)
 
 
+def det_lock_path(lock_name):
+    """
+    Determine full path for lock with specifed name.
+    """
+    locks_dir = build_option('locks_dir') or os.path.join(install_path('software'), '.locks')
+    return os.path.join(locks_dir, lock_name + '.lock')
+
+
+def create_lock(lock_name):
+    """Create lock with specified name."""
+
+    lock_path = det_lock_path(lock_name)
+    _log.info("Creating lock at %s...", lock_path)
+    try:
+        # we use a directory as a lock, since that's atomically created
+        mkdir(lock_path, parents=True)
+        global_lock_names.add(lock_name)
+    except EasyBuildError as err:
+        # clean up the error message a bit, get rid of the "Failed to create directory" part + quotes
+        stripped_err = str(err).split(':', 1)[1].strip().replace("'", '').replace('"', '')
+        raise EasyBuildError("Failed to create lock %s: %s", lock_path, stripped_err)
+    _log.info("Lock created: %s", lock_path)
+
+
+def check_lock(lock_name):
+    """
+    Check whether a lock with specified name already exists.
+
+    If it exists, either wait until it's released, or raise an error
+    (depending on --wait-on-lock configuration option).
+    """
+    lock_path = det_lock_path(lock_name)
+    if os.path.exists(lock_path):
+        _log.info("Lock %s exists!", lock_path)
+        wait_on_lock = build_option('wait_on_lock')
+        if wait_on_lock:
+            while os.path.exists(lock_path):
+                print_msg("lock %s exists, waiting %d seconds..." % (lock_path, wait_on_lock),
+                          silent=build_option('silent'))
+                time.sleep(wait_on_lock)
```

Comment:
I would expect wait_on_lock to be the max time to wait, not how long between checks. That's what you would normally do in this situation, i.e. specify a max time to wait.

Reply:
Yeah, we need a way to limit the time you're waiting for a lock, and the semantics of `--wait-on-lock` is a bit weird right now.

So, I changed things in d879cdadc:

* `--wait-on-lock` still works as before (`0` (the default) implies no waiting (raise an error if lock is already there), a value implies enabling waiting with interval of specified number of seconds), but is deprecated;
* `--wait-on-lock-limit` can be used to i) enable waiting for lock to be released, ii) specify a maximum waiting time (if set to `0` is implies no waiting (raise error if lock is there), if set to `-1` it implies waiting indefinitely);
* `--wait-on-lock-interval` can be used to specify the interval for re-checking whether the lock has been released (default 60 sec.)